tactics/tactics.ml

```diff
@@ -2962,7 +2963,8 @@ let specialize (c,lbind) ipat =
     let sigma = Specialize.unify_bindings sigma unify typ_of_c lbind in
     (* Instantiate unsolved holes with their default value *)
     let fold sigma (env, ev) =
-      if isEvar sigma ev then Evarconv.unify env sigma CONV ev (mkRel 1)
+      if isEvar sigma ev && (fst (destEvar sigma ev)) = (fst (destEvar freeezed_sigma ev))
```

Comment:
The way this test is written is quite weird. You should pattern-match on the kind of the term rather than using a pair isEvar + destEvar. Also, you should probably check that the evar is undefined in freezed_sigma rather than doing the same weird dance with the term expansion. That is, something along the lines of
```ocaml
match EConstr.kind sigma ev with
| Evar (evk, _) when Evd.is_is_undefined freezed_evar evk -> ...
| _ -> ...
```
```

Reply:
`sigma` is supposed to be an extension of `freezed_sigma` so if the evar is undefined in `sigma` it should also be undefined in `freezed_sigma` no?

Reply:
It could also be an evar that was introduced in sigma and that the previous evar was instantiated with through an evar-evar unification. In any case this test needs some comment to assess what it's trying to achieve.

Reply:
The test considers that a hole that has changed during unification should be considered as "instantiated" and not re-quantified as a premise of `H`.  But I get @ppedrot's point: this is too aggressive (as it is for `apply`). We need a middle ground that reject the `with (1:=eq_refl)` but accepts `with (1:= term_with_evars`. I will propose something in the main thread.